source/gameengine/GameLogic/Joystick/SCA_Joystick.cpp

```diff
@@ -200,119 +185,151 @@ bool SCA_Joystick::aAxisIsPositive(int axis_single)
+bool SCA_Joystick::CreateJoystickDevice(void)
 {
-	return (GetHat(hatnum)==dir) ? true : false;
-}
+	bool joy_error = false;
 
-int SCA_Joystick::GetNumberOfAxes()
-{
-	return m_axismax;
-}
+#ifndef WITH_SDL
+	m_isinit = true;
+	joy_error = true;
+#else /* WITH_SDL */
+	if (!m_isinit) {
 
+		if (!joy_error &&
+		    !(SDL_CHECK(SDL_IsGameController) &&
+		    SDL_CHECK(SDL_GameControllerOpen) &&
+		    SDL_CHECK(SDL_GameControllerEventState)) &&
+		    !SDL_IsGameController(m_joyindex))
+		{
+			/* mapping instruccions if joystick is not a game controller */
+			printf("Game Controller index %i: Could not be initialized\n", m_joyindex);
+			printf("Please, generate Xbox360 compatible mapping using antimicro or Steam big mode application\n");
+			printf("and after set, the SDL controller variable before you launch the executable, i.e:\n");
+			printf("export SDL_GAMECONTROLLERCONFIG=\"[the string you received from controllermap]\"\n");
+			/* Need this so python args can return empty lists */
+			joy_error = true;
+		}
 
-int SCA_Joystick::GetNumberOfButtons()
-{
-	return m_buttonmax;
-}
+		if (!joy_error) {
+			m_private->m_gamecontroller = SDL_GameControllerOpen(m_joyindex);
+			if (!m_private->m_gamecontroller) {
+				joy_error = true;
+			}
+		}
 
+		SDL_Joystick *joy;
+		if (!joy_error) {
+			joy = SDL_GameControllerGetJoystick(m_private->m_gamecontroller);
+			if (!joy) {
+				joy_error = true;
+			}
+		}
 
-int SCA_Joystick::GetNumberOfHats()
-{
-	return m_hatmax;
-}
+		if (!joy_error) {
+			m_private->m_instance_id = SDL_JoystickInstanceID(joy);
+			if (m_private->m_instance_id < 0){
+				joy_error = true;
+				printf("Joystick instanced failed: %s\n", SDL_GetError());
+			}
+		}
 
-bool SCA_Joystick::CreateJoystickDevice(void)
-{
-#ifndef WITH_SDL
-	m_isinit = true;
-	m_axismax = m_buttonmax = m_hatmax = 0;
-	return false;
-#else /* WITH_SDL */
-	if (m_isinit == false && SDL_CHECK(SDL_JoystickOpen)) {
-		if (m_joyindex>=m_joynum) {
-			/* don't print a message, because this is done anyway */
-			//JOYSTICK_ECHO("Joystick-Error: " << SDL_NumJoysticks() << " avaiable joystick(s)");
-			
-			/* Need this so python args can return empty lists */
-			m_axismax = m_buttonmax = m_hatmax = 0;
-			return false;
+		if (!joy_error) {
+			printf("\nGame Controller (%s) with index %i: Initialized", GetName(), m_joyindex);
+
+			/* A Game Controller has:
+			 *
+			 * 6 axis availables:	   AXIS_LEFTSTICK_X, AXIS_LEFTSTICK_Y,
+			 * (in order from 0 to 5)  AXIS_RIGHTSTICK_X, AXIS_RIGHTSTICK_Y,
+			 *						   AXIS_TRIGGERLEFT and AXIS_TRIGGERRIGHT.
+			 *
+			 * 15 buttons availables:  BUTTON_A, BUTTON_B, BUTTON_X, BUTTON_Y,
+			 * (in order from 0 to 14) BUTTON_BACK, BUTTON_GUIDE, BUTTON_START,
+			 *						   BUTTON_LEFTSTICK, BUTTON_RIGHTSTICK,
+			 *						   BUTTON_LEFTSHOULDER, BUTTON_RIGHTSHOULDER,
+			 *						   BUTTON_DPAD_UP, BUTTON_DPAD_DOWN,
+			 *						   BUTTON_DPAD_LEFT and BUTTON_DPAD_RIGHT.
+			 */
+			m_axismax = SDL_CONTROLLER_AXIS_MAX;
```

Comment:
`m_axismax` and `m_buttonmax` are constants, no ? I saw later that you set then to 0. These values are used to know if a joystick is valid, or they can be replaced by static function in `SCA_Joystick` ?